src/main/java/net/neoforged/neoforge/client/util/TooltipUtil.java

```diff
@@ -0,0 +1,349 @@
+     */
+    public static void addAttributeTooltips(ItemStack stack, Consumer<Component> tooltip, AttributeTooltipContext ctx) {
+        ItemAttributeModifiers modifiers = stack.getOrDefault(DataComponents.ATTRIBUTE_MODIFIERS, ItemAttributeModifiers.EMPTY);
+        if (modifiers.showInTooltip()) {
+            applyModifierTooltips(stack, tooltip, ctx);
+        }
+        NeoForge.EVENT_BUS.post(new AddAttributeTooltipsEvent(stack, tooltip, ctx));
+    }
+
+    /**
+     * Unconditionally applies the attribute modifier tooltips for all attribute modifiers present on the item stack.
+     * <p>
+     * Before application, this method posts the {@link GatherSkippedAttributeTooltipsEvent} to determine which tooltips should be skipped.
+     * 
+     * @param stack
+     * @param tooltip
+     * @param ctx
+     */
+    public static void applyModifierTooltips(ItemStack stack, Consumer<Component> tooltip, AttributeTooltipContext ctx) {
+        Set<ResourceLocation> skips = new HashSet<>();
+        var event = NeoForge.EVENT_BUS.post(new GatherSkippedAttributeTooltipsEvent(stack, skips, ctx));
+        if (event.isSkippingAll()) {
+            return;
+        }
+
+        for (EquipmentSlotGroup group : EquipmentSlotGroup.values()) {
+            Multimap<Holder<Attribute>, AttributeModifier> modifiers = AttributeUtil.getSortedModifiers(stack, group);
+            Component groupName = Component.translatable("item.modifiers." + group.getSerializedName()).withStyle(ChatFormatting.GRAY);
+            applyTextFor(stack, tooltip, modifiers, groupName, skips, ctx);
+        }
+    }
+
+    /**
+     * Applies the text for a single group of modifiers to the tooltip for a given item stack.
+     * <p>
+     * This method will attempt to merge multiple modifiers for a single attribute into a single modifier if {@linkplain NeoForgeMod#enableMergedAttributeTooltips()} was called.
+     * 
+     * @param stack       The item stack that owns the modifiers
+     * @param tooltip     The consumer to append tooltip components to
+     * @param modifierMap A mutable map of modifiers for the given group
+     * @param groupName   The name of the current modifier group
+     * @param skips       A set of modifier IDs to not apply to the tooltip
+     * @param ctx         The tooltip context
+     */
+    public static void applyTextFor(ItemStack stack, Consumer<Component> tooltip, Multimap<Holder<Attribute>, AttributeModifier> modifierMap, Component groupName, Set<ResourceLocation> skips, AttributeTooltipContext ctx) {
+        // Remove any skipped modifiers before doing any logic
+        modifierMap.values().removeIf(m -> skips.contains(m.id()));
+
+        // Don't add anything if there is nothing in the group
+        if (modifierMap.isEmpty()) {
+            return;
+        }
+
+        // Add an empty line, then the name of the group
+        tooltip.accept(Component.empty());
+        tooltip.accept(groupName);
```

Comment:
For my handling of attribute tooltips; I need the ability to handle all attribute for a specific slot without adding newlines or a groupname. Is there a way this logic could be moved to the above method or possibly add a method between `applyModifierTooltips` and `applyTextFor` so I am able to not have a group name?

Reply:
There's two options here:
1. Move both lines upwards one method, so that group name appending is entirely outside of this method
2. Skip appending the empty line and the group name of the group name is empty (i.e. `== CommonComponents.EMPTY`)

Reply:
Either solution sounds fine to me.

Reply:
I support option 1

Reply:
Option 1 ended up being the better call here.  I first explored option 2, because it's quick to implement, but option 1 produces a better interface by elevating both the `skips` and `groupName` logic to `applyModifierTooltips`. This leaves `applyTextFor` as a "pure" method just to convert the given modifier map into tooltip lines.